docs/proposals/parameter-distribution.md

```diff
@@ -0,0 +1,128 @@
+# Proposal for Parameter Distribution
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Design Details](#design-details)
+  - [Experiment API changes](#experiment-api-changes)
+  - [Correspondence for Katib Distributions and Framework Distributions](#correspondence-for-katib-distributions-and-framework-distributions)
+    - [Chocolate](#chocolate)
+    - [Goptuna](#goptuna)
+    - [Hyperopt](#hyperopt)
+    - [Optuna](#optuna)
+    - [Scikit-Optimize](#scikit-optimize)
+<!-- /toc -->
+
+## Summary
+This enhancement introduces `Distribution` to tuning parameters and remove redundantly `ParameterType`. 
+
+API field in the Experiment spec determine parameter type with distribution.  
+
+## Motivation
+Currently, Katib does not support determining a distribution for search space that samplers pick up parameters by users.
+
+Katib should be able to determine it by users since
+almost hyperparameter tuning algorithms (framework) can determine it by users.
+
+## Proposal
+We introduce a mechanism to determine a distribution for search space by users.
+That also means we introduce a mechanism to propagate distributions to suggestion-services and
+set them samplers of each framework.     
+
+## Design Details
+The proposal consists of a new Experiment API field and
+correspondence between Katib Distributions and Framework Distributions.
+
+### Experiment API changes
+We extend the Experiment API to introduce the new fields `Distribution` to configure the distribution for search space and
+remove the redundant fields `ParameterType`.
+
+```diff
+type ParameterSpec struct {
+	Name          string        `json:"name,omitempty"`
+- 	ParameterType ParameterType `json:"parameterType,omitempty"`
++ 	Distribution  Distribution  `json:"distribution,omitempty"`
+	FeasibleSpace FeasibleSpace `json:"feasibleSpace,omitempty"`
+}
+
+- type ParameterType string
++ type Distribution string
+
+const (
+- 	ParameterTypeUnknown     ParameterType = "unknown"
+- 	ParameterTypeDouble      ParameterType = "double"
+- 	ParameterTypeInt         ParameterType = "int"
+- 	ParameterTypeDiscrete    ParameterType = "discrete"
+- 	ParameterTypeCategorical ParameterType = "categorical"
++ 	UnknownDistribution         Distribution = "unknown"
++ 	CategoricalDistribution     Distribution = "categorical"
++ 	IntUniformDistribution      Distribution = "intUniform"
++ 	IntLogUniformDistribution   Distribution = "intLogUniform"
++ 	FloatUniformDistribution    Distribution = "floatUniform"
++ 	FloatLogUniformDistribution Distribution = "floatLogUniform"
```

Comment:
> What do you think about following [hyperopt model](http://hyperopt.github.io/hyperopt/getting-started/search_spaces/#parameter-expressions) instead of int and float model (e.g. uniform, quniform, loguniform, qloguniform) ? From my point of view, it sounds more native to HP tuning and many HPs papers mention that distribution.

@andreyvelich Sounds good. I would add the corresponding tables for the old ParameterType and new Distribution using the hyperopt model to this proposal.

> Also, we can change step to q and integrate base parameter for the log.

@andreyvelich Sounds good. One question, Does `integrate base parameter for the log` mean adding the `base` field to struct `FeasibleSpace`?

Reply:
SGTM. While I am thinking if this is a huge change to our YAML APIs.

Reply:
> SGTM. While I am thinking if this is a huge change to our YAML APIs.

Maybe, we need to change the API version to v1beta2.

Reply:
SGTM

Reply:
@andreyvelich @johnugeorge @anencore94 wdyt?

Reply:
If we move the API version to v1beta2, it will be helpful to use the [conversion](https://book.kubebuilder.io/multiversion-tutorial/conversion.html).

Reply:
>  Does integrate base parameter for the log mean adding the base field to struct FeasibleSpace

@tenzen-y Yes, I was thinking we can add `base` to the `FeasibleSpace` in the future.

Reply:
@tenzen-y if we are going to create `v1beta2` APIs, it might be hard for us to maintain 2 versions.

Previously, we had to create each component folder for every supported version `v1alpha2` `v1beta1` (e.g. [Suggestion](https://github.com/kubeflow/katib/tree/master/pkg/suggestion) or [Katib DB](https://github.com/kubeflow/katib/tree/master/pkg/db)) 
With the conversion webhook, can we keep 1 version of all components (DB Manager, Controller, UI, Suggestions, etc.) and just keep 2 separate folder for our CRD APIs: https://github.com/kubeflow/katib/tree/master/pkg/apis/controller/experiments ?

Reply:
> With the conversion webhook, can we keep 1 version of all components (DB Manager, Controller, UI, Suggestions, etc.) and just keep 2 separate folder for our CRD APIs:

@andreyvelich IIUC, yes, just maintain multiple CRD versions and one controller version like the cluster-api provider.

https://github.com/kubernetes-sigs/cluster-api-provider-gcp

This means we will maintain the CRDs of v1beta1 and v1beta2, v1beta2 DB Manager, v1beta2 Controller, v1beta2 UI, and v1beta2 Suggestions.

Does that make sense to you?

Reply:
If you have any concerns, please let me know.

Reply:
> With the conversion webhook, can we keep 1 version of all components (DB Manager, Controller, UI, Suggestions, etc.) and just keep 2 separate folder for our CRD APIs:

Yep, I think so. But we need a detailed design for this to see if it is possible. The conversion webhook converts older API versions to newer one, then use the newer controller implementation to deal with all requests. So the question to us is:

- Is it possible to convert v1beta1 resource object to v1beta2? Will it drop some necessary info from the conversion?
- When will the webhook be configured? Should we install it by default? And when will we deprecate v1beta1?

Reply:
SGTM

> Is it possible to convert v1beta1 resource object to v1beta2? Will it drop some necessary info from the conversion?

I will create a correspondence table between v1beta1 and v1beta2. Maybe, we only need to create a table for the `ParameterType` and the `FeasibleSpace`.

> When will the webhook be configured? Should we install it by default? 

IIUC, we do not need to install manifests for conversion webhook to clusters.

ref: 
- https://book.kubebuilder.io/multiversion-tutorial/tutorial.html
- https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/multiversion-tutorial/testdata/project

> And when will we deprecate v1beta1?

IMO, we need to keep maintaining v1beta1 for at least one release version. This means if we introduce v1beta2 API in katib v0.16.0, we will remove v1beta1 API in katib v0.17.0.

@gaocegege Do you know how many release versions we kept maintaining v1alpha2 after we introduced v1beta1?

Reply:
> Yep, I think so. But we need a detailed design for this to see if it is possible.

@andreyvelich @gaocegege  Maybe, custom (implemented by user) suggestion services using v1beta1 API will not work since gRPC calls are not through conversion webhook.

So, we probably need to separate CRD version changes from Distribution introduces. And then I take up only `Introducing Distribution` in this proposal. We can follow up on `Upgrading CRD version` in other issues and PRs.

- Introducing `Distribution`: we keep using `ParameterType` and introducing `Distribution` and `Base` to `FeasibleSpace` like the following.

https://github.com/kubeflow/katib/issues/1207#issuecomment-639275103

- Upgrading CRD version: 

 1. introduce a new field that represents the gRPC API version (v1beta1 or v1beta2) to the following of katib-config since the suggestion controller needs to use a different gRPC client for v1beta1 and v1beta2. This means we keep maintaining both v1beta1 and v1beta2 gRPC APIs (proto) for a while (only gRPC API, no maintaining v1beta1 controller). And then after we remove the v1beta1 API, remove the new field in katib-config.

https://github.com/kubeflow/katib/blob/db72ce1987b11cac09d512587261187465ab2729/pkg/util/v1beta1/katibconfig/config.go#L35-L45

 2. Consolidate `ParameterType` and `FeasibleSpace.Distribution` to `Distribution` using the hyperopt model like @andreyvelich mentioned at https://github.com/kubeflow/katib/pull/2059#discussion_r1046376926.

> @johnugeorge @tenzen-y @gaocegege @anencore94 What do you think about following [hyperopt model](http://hyperopt.github.io/hyperopt/getting-started/search_spaces/#parameter-expressions) instead of int and float model (e.g. uniform, quniform, loguniform, qloguniform) ? From my point of view, it sounds more native to HP tuning and many HPs papers mention that distribution.
Also, we can change step to q and integrate base parameter for the log.
Many data scientists who do HP tuning are familiar with Hyperopt, so the API will look the same for them.
>
> Also, Ray Tune follows the same model: https://docs.ray.io/en/latest/tune/api_docs/search_space.html, and NNI has the same APIs: https://nni.readthedocs.io/en/stable/hpo/search_space.html#quniform


@andreyvelich @gaocegege @johnugeorge @anencore94 wdyt?

Reply:
Hmm, gRPC might be a problem, yes.
Do we know how Kubernetes maintain 2 version of their gRPC APIs ?
e.g. [v1 version for apps](https://github.com/kubernetes/api/blob/master/apps/v1/generated.proto) and [v1beta2 version for apps](https://github.com/kubernetes/api/blob/master/apps/v1beta2/generated.proto) ?

Reply:
@tenzen-y Also, are we going to rename `intuniform` to `quniform` and `floatuniform` to `uniform` as I proposed ?

Reply:
> Do we know how Kubernetes maintain 2 version of their gRPC APIs ?

@andreyvelich  Kubernetes uses helper functions to convert multiple APIs.

https://github.com/kubernetes/kubernetes/blob/c1c0e4fe0bb4e7c0145d45a010577ed64619903a/pkg/apis/apps/v1beta2/conversion.go

Does that answer your question?

> Also, are we going to rename intuniform to quniform and floatuniform to uniform as I proposed ?

Yes, I updated the above comment.